### PR TITLE
Pull requests

### PR DESCRIPTION
**Pull request by wangtzh**
but the zipfian stuff were already there...
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41803191%22%2C%20%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41804022%22%2C%20%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41805147%22%2C%20%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41805662%22%2C%20%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41806087%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%20c6240553578d8dd34d5ea4065a44a53ef80c7142%20foedus-core/include/foedus/xct/xct_id.hpp%2031%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41806087%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Let%27s%20make%20this%2016%20bytes%20in%20near%20future.%5Cr%5Cn%5Cr%5Cnuint32_t%20for%20successor_thread_id%20%28maybe%20you%20want%20to%20embed%20successor_class%20here%29%5Cr%5Cnuint32_t%20for%20successor_block_index%5Cr%5Cnuint64_t%20other-stuffs%5Cr%5Cn%5Cr%5CnFirst%208%20bytes%20are%20written%20only%20by%20successor.%5Cr%5CnSecond%208%20bytes%20needs%20a%20bit%20more%20thinking%20due%20for%20RW-MCSg.%22%2C%20%22created_at%22%3A%20%222015-10-12T22%3A01%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/702382%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/hkimura%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20foedus-core/include/foedus/xct/xct_id.hpp%3AL165-278%22%7D%2C%20%22Pull%20c6240553578d8dd34d5ea4065a44a53ef80c7142%20foedus-core/src/foedus/xct/xct_id.cpp%20125%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41803191%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22readers_count%20and%20next_writer%20field.%5Cr%5Cnthis%20is%20debug-write%2C%20no%20worry%20about%20performance.%20so%20you%20can%20just%20use%20seq-cst%20load%20here.%22%2C%20%22created_at%22%3A%20%222015-10-12T21%3A26%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/702382%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/hkimura%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20foedus-core/src/foedus/xct/xct_id.cpp%3AL324-340%22%7D%2C%20%22Pull%20c6240553578d8dd34d5ea4065a44a53ef80c7142%20tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp%20151%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/hkimura/foedus_code/pull/106%23discussion_r41804022%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Umm%2C%20this%20doesn%27t%20cause%20deadlocks%3F%22%2C%20%22created_at%22%3A%20%222015-10-12T21%3A36%3A09Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/702382%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/hkimura%22%7D%7D%2C%20%7B%22body%22%3A%20%22eh%3F%20it%27s%20released%20immediately%20after%20acquiring%20too%20here.%22%2C%20%22created_at%22%3A%20%222015-10-12T21%3A49%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/9100252%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/wangtzh%22%7D%7D%2C%20%7B%22body%22%3A%20%22Oops%2C%20I%20missed%20that%20line.%20okay%2C%20then%20this%20is%20safe.%22%2C%20%22created_at%22%3A%20%222015-10-12T21%3A56%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/702382%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/hkimura%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp%3AL1-337%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull c6240553578d8dd34d5ea4065a44a53ef80c7142 foedus-core/src/foedus/xct/xct_id.cpp 125'></a> &nbsp;<b><a href='https://github.com/hkimura/foedus_code/pull/106#discussion_r41803191'>File: foedus-core/src/foedus/xct/xct_id.cpp:L324-340</a></b>
- <a href='https://github.com/hkimura'></a> readers_count and next_writer field.
this is debug-write, no worry about performance. so you can just use seq-cst load here.
- [ ] <a href='#crh-comment-Pull c6240553578d8dd34d5ea4065a44a53ef80c7142 tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp 151'></a> &nbsp;<b><a href='https://github.com/hkimura/foedus_code/pull/106#discussion_r41804022'>File: tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp:L1-337</a></b>
- <a href='https://github.com/hkimura'></a> Umm, this doesn't cause deadlocks?
- <a href='https://github.com/wangtzh'></a> eh? it's released immediately after acquiring too here.
- <a href='https://github.com/hkimura'></a> Oops, I missed that line. okay, then this is safe.
- [ ] <a href='#crh-comment-Pull c6240553578d8dd34d5ea4065a44a53ef80c7142 foedus-core/include/foedus/xct/xct_id.hpp 31'></a> &nbsp;<b><a href='https://github.com/hkimura/foedus_code/pull/106#discussion_r41806087'>File: foedus-core/include/foedus/xct/xct_id.hpp:L165-278</a></b>
- <a href='https://github.com/hkimura'></a> Let's make this 16 bytes in near future.
uint32_t for successor_thread_id (maybe you want to embed successor_class here)
uint32_t for successor_block_index
uint64_t other-stuffs
First 8 bytes are written only by successor.
Second 8 bytes needs a bit more thinking due for RW-MCSg.

<a href='https://www.codereviewhub.com/hkimura/foedus_code/pull/106?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/hkimura/foedus_code/pull/106?approve=1'></a>&nbsp;<a href='https://github.com/hkimura/foedus_code/pull/106'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by hkimura** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/hkimura/foedus_code/pull/106'>Refresh</a>.<a href='#crh-update-none'></a>

File: foedus-core/src/foedus/xct/xct_id.cpp [hunk 1]
```diff
@@ -212,5 +324,16 @@ std::ostream& operator<<(std::ostream& o, const LockableXctId& v) {
   return o;
 }
 
+std::ostream& operator<<(std::ostream& o, const McsRwLock& v) {
+  o << "<McsRwLock><locked>" << v.is_locked() << "</locked><tail_waiter>"
+    << v.get_tail_waiter() << "</tail_waiter><tail_block>" << v.get_tail_waiter_block()
```
Comment (hkimura) [inline foedus-core/src/foedus/xct/xct_id.cpp]: readers_count and next_writer field.
this is debug-write, no worry about performance. so you can just use seq-cst load here.

File: tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    McsBlockIndex block = 0;
+    if (id % 2 == 0) {
+      block = context->mcs_acquire_reader_lock(keys[k].get_key_lock());
+      context->mcs_release_reader_lock(keys[k].get_key_lock(), block);
+    } else {
+      block = context->mcs_acquire_writer_lock(keys[k].get_key_lock());
+      context->mcs_release_writer_lock(keys[k].get_key_lock(), block);
+    }
+  }
+  WRAP_ERROR_CODE(xct_manager->abort_xct(context));
```
Comment (hkimura) [inline tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp]: Umm, this doesn't cause deadlocks?

Comment (wangtzh) [inline tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp]: eh? it's released immediately after acquiring too here.

Comment (hkimura) [inline tests-core/src/foedus/xct/test_xct_id_rw_lock.cpp]: Oops, I missed that line. okay, then this is safe.

File: foedus-core/include/foedus/xct/xct_id.hpp [hunk 3]
```diff
@@ -165,6 +165,113 @@ struct McsBlock {
...
+      // Bit 7: blocked (am I waiting for the lock or acquired?)
+      uint8_t state_;
+    } components_;
+  } self_;
+  // TODO(tzwang): make these two fields 8 bytes by themselves. Now we need
+  // to worry about sub-word writes (ie have to use atomic ops even when
+  // changing only these two fields because they are in the same byte as data_).
+  thread::ThreadId successor_thread_id_;  // +2 => 4
+  McsBlockIndex successor_block_index_;   // +4 => 8
+
```
Comment (hkimura) [inline foedus-core/include/foedus/xct/xct_id.hpp]: Let's make this 16 bytes in near future.

uint32_t for successor_thread_id (maybe you want to embed successor_class here)
uint32_t for successor_block_index
uint64_t other-stuffs

First 8 bytes are written only by successor.
Second 8 bytes needs a bit more thinking due for RW-MCSg.

**Comment by hkimura** [comment]
Oops, could you run the following?
./build/tests-core/src/foedus/test_headers --gtest_filter=*.IndependenceCXX98
I think it complains.

**Pull request merged**